crates/accelerate/src/gate_direction.rs

```diff
@@ -0,0 +1,159 @@
+#[pyfunction]
+#[pyo3(name = "check_gate_direction_coupling")]
+fn py_check_with_coupling_map(
+    py: Python,
+    dag: &DAGCircuit,
+    coupling_edges: &Bound<PySet>,
+) -> PyResult<bool> {
+    let coupling_map_check =
+        |curr_dag: &DAGCircuit, _: &PackedInstruction, op_args: &[Qubit]| -> bool {
+            coupling_edges
+                .contains((
+                    map_qubit(py, dag, curr_dag, op_args[0]).0,
+                    map_qubit(py, dag, curr_dag, op_args[1]).0,
+                ))
+                .unwrap_or(false)
+        };
+
+    check_gate_direction(py, dag, &coupling_map_check)
+}
+
+/// Check if the two-qubit gates follow the right direction with respect to instructions supported in the given target.
+///
+/// Args:
+///     dag: the DAGCircuit to analyze
+///
+///     target: the Target against which gate directionality compliance is checked
+///
+/// Returns:
+///     true iff all two-qubit gates comply with the target's coupling constraints
+#[pyfunction]
+#[pyo3(name = "check_gate_direction_target")]
+fn py_check_with_target(py: Python, dag: &DAGCircuit, target: &Bound<Target>) -> PyResult<bool> {
+    let target = target.borrow();
+
+    let target_check =
+        |curr_dag: &DAGCircuit, inst: &PackedInstruction, op_args: &[Qubit]| -> bool {
+            let mut qargs = Qargs::new();
+
+            qargs.push(PhysicalQubit::new(
+                map_qubit(py, dag, curr_dag, op_args[0]).0,
+            ));
+            qargs.push(PhysicalQubit::new(
+                map_qubit(py, dag, curr_dag, op_args[1]).0,
+            ));
+
+            target.instruction_supported(inst.op.name(), Some(&qargs))
+        };
+
+    check_gate_direction(py, dag, &target_check)
+}
+
+// Handle a control flow instruction, namely check recursively into its circuit blocks
+fn check_gate_direction_control_flow<T>(
+    py: Python,
+    py_inst: &PyInstruction,
+    gate_complies: &T,
+) -> PyResult<bool>
+where
+    T: Fn(&DAGCircuit, &PackedInstruction, &[Qubit]) -> bool,
+{
+    let circuit_to_dag = imports::CIRCUIT_TO_DAG.get_bound(py); // TODO: Take out of the recursion
+    let py_inst = py_inst.instruction.bind(py);
+
+    let raw_blocks = py_inst.getattr("blocks")?;
+    let blocks = raw_blocks.downcast::<PyTuple>()?;
+
+    for block in blocks.iter() {
+        let inner_dag: DAGCircuit = circuit_to_dag.call1((block,))?.extract()?;
+
+        if !check_gate_direction(py, &inner_dag, gate_complies)? {
+            return Ok(false);
+        }
+    }
+
+    Ok(true)
+}
+
+// The main routine for checking gate directionality
+fn check_gate_direction<T>(py: Python, dag: &DAGCircuit, gate_complies: &T) -> PyResult<bool>
+where
+    T: Fn(&DAGCircuit, &PackedInstruction, &[Qubit]) -> bool,
+{
+    for node in dag.op_nodes(false) {
+        let NodeType::Operation(packed_inst) = &dag.dag[node] else {
+            return Err(DAGCircuitError::new_err("PackedInstruction is expected"));
```

Comment:
This gets returned to python and a python user doesn't know what `PackedInstruction` is. I'd say a `panic!()` here is actually probably fine because it's a rust domain bug if we get a non-operation here as `op_nodes` should be filtering it for us. But that being said I'd say if you did want a Python error return I'd change this to be:

```suggestion
            return Err(DAGCircuitError::new_err("An operation node was expected"));
```

Reply:
You're right. Switched to using `panic!`